### PR TITLE
Extension redesign

### PR DESCRIPTION
**Pull request by qubixes**
This also breaks backward compatibility with the GMF file.

Changes:

- dist_tree or equivalents are renamed to distribution_provider
- According to that the name of the entry-point has also changed from `metasynth.disttree` to `metasynth.distribution_provider`. Thus, after switching to the new version/branch, you should re-install metasynth.
- Add a routine to test the validation of the GMF file for plugins. `check_distribution_validation(provider_name)`
- New module `metasynth.validation` that deals with creating validation schemas and checking them dynamically (or statically).
- Changed `FakerDistribution`, and removed the code for using the `faker.city`. Instead, users should import the FakerDistribution and use `FakerDistribution("city"). This makes our code more consistent at a small price.
- Add the new class structures that derive from `metasynth.privacy.BasePrivacy`. Instead of using a string to set the level/kind of privacy for `MetaDataset.from_dataframe`, they should now use the appropriate class from the privacy package they want to use. For example `DisclosurePrivacy` or `DifferentialPrivacy`, with the appropriate parameters (e.g. epsilon). The privacy can now also be set per column with the spec.
- For distributions, the following class variables (should be all lower case) have changed:
   - aliases -> removed, now only one name/identifier is used.
   - implements -> Should include the distribution set (for example `core`), and identifier (for example `uniform`). Re-implementations of distributions, should set it to the same as the original distribution. So, for example if a distribution implements a faster version of the `core.uniform` distribution, it should also have `implements = "core.uniform"`
   - provenance -> Should say which package/plugin provides the implementation.
   - privacy -> Type of de-identification methods that are being used. For example `none`, `disclosure` or `differential_privacy`.
- Schemas are now part of the distribution. Only the parameter part of the schema needs to be set for distributions. It should be a dictionary with the schema for each of the parameters. The method to be implemented is `BaseDistribution._param_schema`
- Instead of implementing `BaseDistribution.to_dict`, distributions are now encouraged to simply implement `BaseDistribution._param_dict`, which should be a dictionary of the (non-privacy) parameters of the distribution.

### CODE REVIEW

File: metasynth/dataset.py [hunk 1]
```diff
@@ -15,7 +15,9 @@
 import jsonschema
 
 from metasynth.var import MetaVar
-from metasynth.disttree import get_disttree
+from metasynth.provider import BaseDistributionProvider
+from metasynth.validation import validate_gmf_dict
+from metasynth.privacy import BasePrivacy, NoPrivacy
```
Comment (J535D165) [inline metasynth/dataset.py:20]: It might be nice to use `isort` for sorting imports.

**Review by J535D165 — COMMENTED** [review]
Some nice changes!

File: metasynth/dataset.py (hunk 1 above)
Comment (qubixes) [reply metasynth/dataset.py]: Thanks for the feedback, should be fixed now!

File: metasynth/distribution/base.py [hunk 2]
```diff
@@ -70,8 +72,40 @@ def __str__(self) -> str:
...
+    def _param_dict(self):
+        """Get dictionary with the parameters of the distribution."""
+
+    def to_dict(self) -> dict:
         """Convert the distribution to a dictionary."""
+        return {
+            "implements": self.implements,
+            "provenance": self.provenance,
+            "class_name": self.__class__.__name__,
+            "parameters": self._param_dict(),
```
Comment (qubixes) [inline metasynth/distribution/base.py:84]: Make deep copy

File: metasynth/provider.py [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+            unique = False
+
+        dist_aic = [dist_aic[i] for i in range(len(dist_aic))
+                    if dist_instances[i].is_unique == unique]
+        dist_instances = [d for d in dist_instances if d.is_unique == unique]
+        if len(dist_instances) == 0:
+            raise ValueError(f"No available distributions for variable '{series.name}'"
+                             f" with variable type '{var_type}' "
+                             f"that have unique == {unique}.")
+        return dist_instances[np.argmin(dist_aic)]
```
Comment (vankesteren) [inline metasynth/provider.py:126]: This is now unused, fitting happens in distributionlist

```suggestion
```

File: metasynth/provider.py [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+
+        Returns
+        -------
+        tuple[Type[BaseDistribution], dict[str, Any]]:
+            A distribution and the arguments to create an instance.
+        """
+        for dist_class in self.distributions:
+            if dist_class.is_named(dist_name) and dist_class.privacy == privacy:
+                return dist_class
+        raise ValueError(f"Cannot find distribution with name '{dist_name}'.")
```
Comment (vankesteren) [inline metasynth/provider.py:157]:
```suggestion
```

This makes `test_distribution.py` fail, should be amended.

File: metasynth/provider.py [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+        elif inspect.isclass(dist) and issubclass(dist, BaseDistribution):
+            dist_instance = dist.fit(series, **fit_kwargs)
+        if isinstance(dist, BaseDistribution):
+            dist_instance = dist
+
+        if dist_instance is None:
+            raise TypeError(
+                f"Distribution with type {type(dist)} is not a BaseDistribution")
+
+        return dist_instance
```
Comment (vankesteren) [inline metasynth/provider.py:194]:
```suggestion
```

File: metasynth/provider.py [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+        BaseDistribution:
+            Distribution representing the dictionary.
+        """
+        for dist_class in self.get_dist_list(var_dict["type"]):
+            if dist_class.implements == var_dict["distribution"]["implements"]:
+                return dist_class.from_dict(var_dict["distribution"])
+        raise ValueError(f"Cannot find distribution with name "
+                         f"'{var_dict['distribution']['implements']}'"
+                         f"and type '{var_dict['type']}'.")
+
```
Comment (vankesteren) [inline metasynth/provider.py:215]: Move this to DistributionList

File: metasynth/provider.py [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+                             f"that have unique == {unique}.")
+        return dist_instances[np.argmin(dist_aic)]
+
+    @property
+    def all_var_types(self) -> List[str]:
+        """Return list of available variable types."""
+        return [
+            "categorical", "discrete", "continuous",
+            "string", "datetime", "date", "time"
+        ]
```
Comment (vankesteren) [inline metasynth/provider.py:134]: Get this from distribution list

File: metasynth/provider.py [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+
+    @property
+    @abstractmethod
+    def name(self):
+        """Name of the distribution package."""
+
+    @property
+    @abstractmethod
+    def version(self):
+        """Version of the distribution package."""
```
Comment (vankesteren) [inline metasynth/provider.py:68]: Make these class attributes and test in __init__ whether they are not None

File: metasynth/provider.py [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+        return [
+            DiscreteUniformDistribution, PoissonDistribution, UniqueKeyDistribution,
+            UniformDistribution, NormalDistribution, LogNormalDistribution,
+            TruncatedNormalDistribution, ExponentialDistribution,
+            MultinoulliDistribution,
+            RegexDistribution, UniqueRegexDistribution, FakerDistribution,
+            UniformDateDistribution,
+            UniformTimeDistribution,
+            UniformDateTimeDistribution,
+        ]
```
Comment (vankesteren) [inline metasynth/provider.py:239]: These will be class attributes

File: metasynth/provider.py [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+
+
+class DistributionProviderList():  # pylint: disable=too-few-public-methods
+    """List of DistributionProvider's with functionality to fit distributions.
```
Comment (vankesteren) [inline metasynth/provider.py:243]:
```suggestion
    """List of DistributionProviders with functionality to fit distributions.
```

File: metasynth/provider.py [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+            if isinstance(pkg, str):
+                self.dist_packages.append(get_distribution_provider(pkg))
+            elif isinstance(pkg, type):
+                self.dist_packages.append(pkg())
+            elif isinstance(pkg, BaseDistributionProvider):
+                self.dist_packages.append(pkg)
+            else:
+                raise ValueError(f"Unknown distribution package type '{type(pkg)}'")
+
+    def fit(self, series: pl.Series,  # pylint: disable=too-many-arguments
```
Comment (vankesteren) [inline metasynth/provider.py:274]: I don't like these comments. Let's configure pylint globally in `pyproject.toml`

File: metasynth/provider.py [hunk 12]
```diff
@@ -0,0 +1,456 @@
...
+            raise ValueError(f"No available distributions for variable '{series.name}'"
+                             f" with variable type '{var_type}' "
+                             f"that have unique == {unique}.")
+        return dist_instances[np.argmin(dist_aic)]
+
+    def _find_distribution(self, dist_name: str, privacy: BasePrivacy = NoPrivacy(),
+                           ) -> type[BaseDistribution]:
+        """Find a distribution and fit keyword arguments from a name.
+
+        This allows us to use 'faker.city' to generate a faker instance that generates cities.
```
Comment (vankesteren) [inline metasynth/provider.py:355]:
```suggestion
```

File: metasynth/distribution/base.py [hunk 13]
```diff
@@ -102,28 +136,12 @@ def is_named(cls, name: str) -> bool:
...
-        return {}
-
-    @property
-    def name(self) -> str:
-        """Return the name used in the metadata file."""
-        return self.aliases[0]
+        assert cls.implements != "unknown", f"Internal error in class {cls.__name__}"
+        return name in (cls.implements.split(".")[1],
+                        cls.implements,
+                        type(cls).__name__,
```
Comment (qubixes) [inline metasynth/distribution/base.py:142]:
```suggestion
```

File: metasynth/provider.py [hunk 14]
```diff
@@ -0,0 +1,456 @@
...
+
+        Parameters
+        ----------
+        dist_name:
+            Name of the distribution, such as faker.city, DiscreteUniformDistribution or normal.
```
Comment (vankesteren) [inline metasynth/provider.py:360]:
```suggestion
            Name of the distribution, e.g., for the built-in uniform distribution: "uniform", "core.uniform", "UniformDistribution".
```

File: metasynth/provider.py [hunk 15]
```diff
@@ -0,0 +1,456 @@
...
+    It has a property {var_type}_distributions for every var_type.
+    """
+
+    def __init__(self, **kwargs):
+        # Perform internal consistency check.
+        for var_type in self.all_var_types:
+            for dist in self.get_dist_list(var_type):
+                assert dist.var_type == var_type, (f"Error: Distribution tree is inconsistent for "
+                                                   f"{dist}.")
+        self.privacy_kwargs = kwargs
```
Comment (qubixes) [inline metasynth/provider.py:53]:
```suggestion
```

File: metasynth/provider.py [hunk 16]
```diff
@@ -0,0 +1,456 @@
...
+        return provider
+    if isinstance(provider, type):
+        return provider()
+
+    all_disttrees = {
+        entry.name: entry
+        for entry in entry_points(group="metasynth.distribution_provider")
+    }
+    try:
+        return all_disttrees[provider].load()(**kwargs)
```
Comment (vankesteren) [inline metasynth/provider.py:454]:
```suggestion
    all_providers = {
        entry.name: entry
        for entry in entry_points(group="metasynth.distribution_provider")
    }
    try:
        return all_providers[provider].load()
```

File: metasynth/provider.py [hunk 17]
```diff
@@ -0,0 +1,456 @@
...
+                dist_list.extend(dist_pkg.distributions)
+            else:
+                dist_list.extend(dist_pkg.get_dist_list(var_type))
+
+        dist_list = [dist for dist in dist_list if dist.privacy == privacy.name]
+        return dist_list
+
+
+def get_distribution_provider(
+        provider: Union[str, type, BaseDistributionProvider] = "builtin", **kwargs
```
Comment (vankesteren) [inline metasynth/provider.py:428]:
```suggestion
        provider: Union[str, type[BaseDistributionProvider], BaseDistributionProvider] = "builtin"
```

File: metasynth/testutils.py [hunk 18]
```diff
@@ -1,46 +1,62 @@
 """Testing utilities for plugins."""
 
-from typing import Optional
+# from typing import Optional
```
Comment (qubixes) [inline metasynth/testutils.py:3]:
```suggestion
```

File: metasynth/testutils.py [hunk 19]
```diff
@@ -1,46 +1,62 @@
...
-        if var_type == "category":
-            series = pl.Series([dist.draw() for _ in range(100)], pl.Categorical)
-        else:
-            series = pl.Series([dist.draw() for _ in range(100)])
-        assert isinstance(new_class.fit(series, **privacy_kwargs), base_class)
-        assert new_tree.fit(series, var_type).var_type == var_type
+    # package = get_distribution_provider(package_name)
+    # for dist in package.distributions:
+    schema = distribution.schema()
+    dist_dict = distribution.default_distribution().to_dict()
```
Comment (vankesteren) [inline metasynth/testutils.py:47]: Is this an abstract method?

File: metasynth/validation.py [hunk 20]
```diff
@@ -0,0 +1,91 @@
+"""Tools for validating distribution/GMF file output."""
```
Comment (vankesteren) [inline metasynth/validation.py:1]: We should have a discussion at a later point about this

File: metasynth/var.py [hunk 21]
```diff
@@ -2,14 +2,17 @@
...
 
 from metasynth.distribution.base import BaseDistribution
-from metasynth.disttree import BaseDistributionTree, get_disttree
+from metasynth.privacy import BasePrivacy, NoPrivacy
+from metasynth.provider import (BaseDistributionProvider,
+                                DistributionProviderList,
+                                get_distribution_provider)
 
 
 def _to_polars(series: Union[pd.Series, pl.Series]) -> pl.Series:
```
Comment (vankesteren) [inline metasynth/var.py:18]: Move to bottom of file

File: metasynth/var.py [hunk 22]
```diff
@@ -185,26 +190,23 @@ def fit(self,
...
-            By default use all distributions in metasynth.distribution.
+        dist_providers:
+            Distribution providers that are used for fitting.
+        privacy:
+            Privacy level to use for fitting the series.
         unique:
             Whether the variable should be unique. If not supplied, it will be
             inferred from the data.
+        fit_kwargs:
+            Extra options for distributions during the fitting stage.
```
Comment (vankesteren) [inline metasynth/var.py:201]: can we have an example of this at some point in the future?

File: metasynth/var.py [hunk 23]
```diff
@@ -185,26 +190,23 @@ def fit(self,
...
-        disttree = get_disttree(distribution_tree)
-
-        # Manually supplied distribution
-        if dist is None:
-            self.distribution = disttree.fit(self.series, self.var_type, unique=unique,
-                                             **fit_kwargs)
-        else:
-            self.distribution = disttree.fit_distribution(dist, self.series, **fit_kwargs)
+        pkg_list = DistributionProviderList(dist_providers)
+        self.distribution = pkg_list.fit(self.series, self.var_type, dist, privacy, unique,
```
Comment (vankesteren) [inline metasynth/var.py:208]:
```suggestion
        provider_list = DistributionProviderList(dist_providers)
        self.distribution = provider_list.fit(self.series, self.var_type, dist, privacy, unique,
```

File: tests/test_distribution.py [hunk 24]
```diff
@@ -6,28 +6,26 @@
...
-        ("faker.city.nl_NL", FakerDistribution),
+        ("faker", FakerDistribution),
         ("regex", RegexDistribution),
         ("discrete_uniform", DiscreteUniformDistribution)
     ]
 )
 def test_util(dist_str, dist):
-    dist_tree = get_disttree()
-    dist_class, _ = dist_tree.find_distribution(dist_str)
+    dist_tree = get_distribution_provider()
```
Comment (vankesteren) [inline tests/test_distribution.py:24]:
```suggestion
    provider_list = DistributionProviderList("builtin")
```

**Review by vankesteren — CHANGES_REQUESTED** [review]
Great work! Huge changes and improvements overall.

File: metasynth/dataset.py [hunk 25]
```diff
@@ -91,25 +92,28 @@ def from_dataframe(cls,
...
 
         Returns
         -------
         MetaDataset:
             Initialized MetaSynth dataset.
         """
-        distribution_tree = get_disttree(privacy_package, **privacy_kwargs)
-
+        if privacy is None:
+            privacy = NoPrivacy()
```
[the remark this answers is not in the document]
Comment (qubixes) [reply metasynth/dataset.py:116]: What about `BasicPrivacy`?

File: metasynth/testutils.py (hunk 19 above)
Comment (qubixes) [reply metasynth/testutils.py:46]: Yes it is.

File: metasynth/dataset.py [hunk 26]
```diff
@@ -122,23 +126,18 @@ def from_dataframe(cls,
...
             unq = col_spec.pop("unique", None)
             description = col_spec.pop("description", None)
-            fit_kwargs = col_spec.pop("fit_kwargs", {})
             prop_missing = col_spec.pop("prop_missing", None)
-            assert "fit_kwargs" not in col_spec
-            if dist is None and len(fit_kwargs) > 0:
-                raise ValueError(f"Got fit arguments for variable '{col_name}', but no "
-                                 "distribution. Set the distribution manually to fix.")
+            cur_privacy = col_spec.pop("privacy", privacy)
+            fit_kwargs = col_spec.pop("fit_kwargs", {})
```
[the remark this answers is not in the document]
Comment (qubixes) [reply metasynth/dataset.py:131]: Changed in ProviderList

**Pull request merged**